### PR TITLE
【OSCP】在 SecretFlow 拆分学习中实现 Equation Solving 攻击

### PR DESCRIPTION
**Pull request by Rainysponge**
**Type of change**
fixed: #1239
- [ ] Add new papers (Please tell us why you think this paper is awesome!)
- [ ] Fix the category of an existing paper/papers (Please tell us the reasons)
- [x] Add a new tool/primitive/application with a new markdown page (Thank you! Also, please tell us more about this awesome thing!)

**Description**

>This PR introduces the [Equation Solving Attack](https://arxiv.org/abs/2210.06771) into SecretFlow. The method is designed for reconstructing binary features using a robust search-based attack.

**Result**
In fact, in my experiments, the variance of the results of this method is large.

94264f26c9b7c07699ae47cda0c58f5

image

### CODE REVIEW

File: tests/ml/nn/sl/attack/test_torch_solving_linear_regression_fia.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        base_model_dict=base_model_dict,
+        device_y=bob,
+        model_fuse=fuse_model,
+        dp_strategy_dict=None,
+        compressor=None,
+        simulation=True,
+        random_seed=1234,
+        backend='torch',
+        strategy='split_nn',
+        agg_method=agg_method,
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/test_torch_solving_linear_regression_fia.py:218]: agg_method 为空的话，fuse_model 可以直接拿到 victim hidden，上面的 get_victim_hidden 可以省了

Comment (zhaocaibei123) [reply tests/ml/nn/sl/attack/test_torch_solving_linear_regression_fia.py:218]: fuse_model 定义 把 x concate 一下，可以参考 secretflow/ml/nn/applications/sl_dnn_torch.py  DnnFuse

File: secretflow/ml/nn/sl/attacks/solving_linear_regression_fia_torch.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+import torch
+
+from secretflow import reveal
+from secretflow.device import PYU
+from secretflow.ml.nn.callbacks.attack import AttackCallback
+
+
+class SolvingLinearRegressionAttack(AttackCallback):
+    """
+    TODO: need explaination
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/attacks/solving_linear_regression_fia_torch.py:27]: 补一下注释

**Review by zhaocaibei123 — APPROVED** [review]

**Review by ian-huu — APPROVED** [review]
LGTM

**Pull request merged**